client-misk-jooq/src/main/kotlin/app/cash/backfila/client/misk/jooq/internal/JooqBackfillOperator.kt

```diff
@@ -0,0 +1,147 @@
+import okio.ByteString
+import java.util.stream.Collectors
+
+/**
+ * @param <K> the type of key the backfill iterates over.
+ * @param <Param> backfill parameters
+ */
+class JooqBackfillOperator<K, Param : Any> internal constructor(
+  override val backfill: JooqBackfill<K, Param>,
+  private val parametersOperator: BackfilaParametersOperator<Param>
+) : BackfillOperator {
+
+  override fun name(): String = backfill.javaClass.name
+
+  override fun prepareBackfill(request: PrepareBackfillRequest): PrepareBackfillResponse {
+    backfill.prepareAndValidateBackfill(
+      parametersOperator.constructBackfillConfig(
+        request.parameters, request.dry_run
+      )
+    )
+
+    val partitions = backfill.shardedTransacterMapBackfill.keys.map { partitionName ->
+      Partition.Builder()
+        .partition_name(partitionName)
+        .backfill_range(computeOverallRange(partitionName, request))
+        .estimated_record_count(null)
+        .build()
+    }
+
+    return PrepareBackfillResponse.Builder().partitions(partitions).build()
+  }
+
+  override fun getNextBatchRange(request: GetNextBatchRangeRequest): GetNextBatchRangeResponse {
+    check(request.compute_count_limit > 0) { "batch limit must be > 0" }
+
+    if (request.backfill_range.start == null) {
+      // This partition never had any data, stop it immediately.
+      return GetNextBatchRangeResponse.Builder()
+        .batches(emptyList())
+        .build()
+    }
+    val config = parametersOperator.constructBackfillConfig(
+      request.parameters, request.dry_run
+    )
+
+    return backfill.inTransactionReturning(
+      "${name()}#JooqBackfillOperator#getNextBatchRange",
+      request.partition_name
+    ) { dslContext ->
+      GetNextBatchRangeResponse.Builder()
+        .batches(
+          Streams.stream(BatchRangeIterator(backfill, dslContext, request, config))
+            .limit(request.compute_count_limit)
+            .collect(Collectors.toList())
+        )
+        .build()
+    }
+  }
+
+  override fun runBatch(request: RunBatchRequest): RunBatchResponse {
+    val config = parametersOperator.constructBackfillConfig(
+      request.parameters, request.dry_run
+    )
+
+    val keysInRange = backfill.inTransactionReturning(
+      "${name()}#JooqBackfillOperator#runBatch", request.partition_name
+    ) { dslContext ->
+      dslContext.select(backfill.compoundKeyFields)
+        .from(backfill.table)
+        .where(backfill.filterCondition(config))
+        .and(
+          backfill.compareCompoundKey(
+            backfill.fromByteString(request.batch_range.start), CompoundKeyComparer<K>::gte
+          )
+        )
+        .and(
+          backfill.compareCompoundKey(
+            backfill.fromByteString(request.batch_range.end), CompoundKeyComparer<K>::lte
+          )
+        )
+        .orderBy(backfill.compoundKeyFields)
+        .fetch { backfill.recordToKey(it) }
+    }
+
+    backfill.backfill(
+      BackfillBatch(
```

Comment:
I wonder if we really need this wrapper or if we can just consider them method parameters?

Reply:
The JooqBackfill.backfill method is something this client controls so I think, yes?

Reply:
I like having the wrapper. If we need to add anything additional to be passed down to the implementors of the backfill, then we can do so easily without breaking any existing implementations.

Reply:
Yeah good point. I think the API is stable now but who knows. We'll see.